Add neo RemoveChannel request

gavin-norman-sociomantic opened the pull request:
Blocked on https://github.com/sociomantic-tsunami/swarm/milestone/27.

Commit:
Remove unneeded imports

Commit:
Remove erroneous comment

Commit:
Add test for RemoveChannel

gavin-norman-sociomantic commented:
A bit WIP. Requires a few doc fixes and more testing.

Commit:
Switch neo test cases to use neo RemoveChannel

gavin-norman-sociomantic commented:
This needs to be reworked to allow RemoveChannel to work on a mirrored channel. See https://github.com/sociomantic-tsunami/dhtproto/pull/61#issuecomment-343895187.

Commit:
Add client, common, and node RemoveChannel protocol

Commit:
Add fakedht RemoveChannel implementation

Commit:
Adapt Mirror test to use neo RemoveChannel

Commit:
Use "admin" client in tests

The tests require the use of the RemoveChannel request. The real DHT node
restricts this to the "admin" client.

gavin-norman-sociomantic commented:
Adapted to not include change to Mirror request. Tested with dhtnode (https://github.com/sociomantic-tsunami/dhtnode/pull/43).

Commit:
Mention RemoveChannel in client README

gavin-norman-sociomantic commented:
Updated: added flush call.

gavin-norman-sociomantic commented:
Updated to use new SupportedCode from swarm v4.5.0.

gavin-norman-sociomantic commented:
Based on #94. Blocked.

gavin-norman-sociomantic commented:
Fixed D2 bug.

gavin-norman-sociomantic commented:
#94 merged. Unblocked.

nemanja-boric-sociomantic commented on src/dhttest/DhtTestCase.d:
```diff
@@ -214,12 +214,10 @@ abstract class NeoDhtTestCase : TestCase
-
-        auto legacy_dht = new Legacy.DhtClient;
-        legacy_dht.handshake(10000);
-        legacy_dht.removeChannel(this.test_channel);
```
Oh, this is a nice consequence!

gavin-norman-sociomantic replied:
Yep!

nemanja-boric-sociomantic commented on src/dhttest/cases/neo/GetChannels.d:
```diff
@@ -62,13 +62,10 @@ public class GetChannels : NeoDhtTestCase
-            legacy_dht.removeChannel(channel);
+            this.dht.blocking.removeChannel(channel);
+
+        super.cleanup();
```
Why is this moved after the `removeChannel`?

gavin-norman-sociomantic replied:
Hmm, not sure if there was a reason for that or if it was a mistake. I can try compiling it without that change.

gavin-norman-sociomantic replied:
Yeah, the reason is that the last thing `super.cleanup` does is shut down the DHT client. Previously, when this test was using a separate client instance to remove the used channels, this had no effect.

nemanja-boric-sociomantic replied:
Makes sense :+1:

nemanja-boric-sociomantic commented on src/dhttest/cases/neo/GetChannels.d:
```diff
@@ -118,12 +113,9 @@ public class GetChannelsRemove : NeoDhtTestCase
-            legacy_dht.removeChannel(channel);
+            this.dht.blocking.removeChannel(channel);
+
+        super.cleanup();
```
Why is this moved after the removeChannel?

nemanja-boric-sociomantic commented on relnotes/removechannel.feature.md:
```diff
@@ -0,0 +1,13 @@
+## Neo RemoveChannel request
```
Use new release notes format.

nemanja-boric-sociomantic commented on src/dhtproto/client/mixins/NeoSupport.d:
```diff
@@ -1430,6 +1466,92 @@ template NeoSupport ( )
+            void notifier ( Neo.RemoveChannel.Notification info,
+                Neo.RemoveChannel.Args args )
+            {
+                with ( info.Active ) switch ( info.active )
```
`final` switch?

Commit:
Add client removeChannel methods (blocking and callback-based)

Fixes #20.

gavin-norman-sociomantic commented:
Updated:
* Release notes. (Including extra commit fixing Exists notes.)
* `final switch`.

gavin-norman-sociomantic commented:
Fixed D2 build.

nemanja-boric-sociomantic commented:
This is red, though, I'm not sure if it's just the matter of a flaky test.

gavin-norman-sociomantic commented:
Huh, it was green before I rebased.

gavin-norman-sociomantic commented:
```
[turtle    ] Testing 'Neo Mirror followed by neo Puts and suspend/resume' ...
[turtle    ]     FAIL at ./src/dhttest/cases/neo/Mirror.d:349
[turtle    ]         (expression '1012 == 1014' evaluates to false)
```
I can reproduce this failure locally. The interesting thing is that this only manifests now after this branch was rebased.

gavin-norman-sociomantic commented:
Looks timing related :( Removing random tests from the test suite makes it pass again.

Commit:
Fix flaky Mirror test by only writing a fixed number of records

Previously, the number of records written during the Mirror test could
vary: the Put timer only stopped once a fixed number of Put requests had
been reported to the client as having succeeded. The problem with this
approach is that Put requests may have been handled by the node and
forwarded to the Mirror request before the client is notified that the Put
has succeeded. This made the test flaky, as it was affected by the timing
of the individual requests.

Now, the test always performs a fixed number of Put requests.

gavin-norman-sociomantic commented:
Updated: added another commit that removes flakiness from the Mirror test.

gavin-norman-sociomantic commented:
Green! :balloon:

nemanja-boric-sociomantic commented on src/dhtproto/client/request/internal/RemoveChannel.d:
```diff
@@ -0,0 +1,324 @@
+    private bool handleSupportedCode ( ubyte code )
+    {
+        auto supported = cast(SupportedStatus)code;
+        if ( !RemoveChannel.handleSupportedCodes(supported,
```
`return RemoveChannel.handleSupportedCodes(...)`

nemanja-boric-sociomantic replied:
Ping ;-)

nemanja-boric-sociomantic commented on src/dhtproto/client/request/internal/RemoveChannel.d:
```diff
@@ -0,0 +1,324 @@
+    Client DHT RemoveChannel v0 request handler.
+
+    Copyright:
+        Copyright (c) 2017 sociomantic labs GmbH. All rights reserved.
```
Copyright 2018.

nemanja-boric-sociomantic replied:
(in other files as well)

gavin-norman-sociomantic replied:
Well, this code was written in 2017 ;)

nemanja-boric-sociomantic replied:
Yeah, I was wondering about that. All right, as long as I have a precedent to refer to in the reviews  I receive ;-)

nemanja-boric-sociomantic commented on src/dhtproto/node/neo/request/RemoveChannel.d:
```diff
@@ -0,0 +1,116 @@
+                payload.add(this.response);
+            }
+        );
+        ed.flush();
```
I guess we can leave the `flush` until we move to the new swarm.

nemanja-boric-sociomantic replied:
Which reminds me, DHT's requests would benefit from `TCP_NODELAY`, even with the current swarm.

nemanja-boric-sociomantic commented on src/dhttest/cases/neo/Mirror.d:
```diff
@@ -549,9 +548,8 @@ public class MirrorRemoveChannel : NeoDhtTestCase
-                            this.dht.removeChannel(this.test_channel,
-                                &delNotifier));
+                        this.dht.neo.removeChannel(this.test_channel,
+                            &delNotifier);
```
Fun fact: this can be `null` ;-) (not that I would recommend that, I think this way it's clearer).

nemanja-boric-sociomantic commented on src/dhtproto/client/mixins/NeoSupport.d:
```diff
@@ -455,6 +457,40 @@ template NeoSupport ( )
+
+            auto id = this.assign!(Internals.RemoveChannel)(params);
+            return id;
+        }
```
Not sure if it's just the github, but while you're there, it would be nice if this commit would be before adapting the mirror tests to use this method.

gavin-norman-sociomantic replied:
Yeah, it is.

nemanja-boric-sociomantic commented on src/dhtproto/client/mixins/NeoSupport.d:
```diff
@@ -1610,6 +1646,92 @@ template NeoSupport ( )
+
+        ***********************************************************************/
+
+        private static struct RemoveChannelResult
```
I forgot, but I think `static` here (in non-function context) is redundant.

gavin-norman-sociomantic replied:
Correct :+1:

nemanja-boric-sociomantic commented on src/dhtproto/client/mixins/NeoSupport.d:
```diff
@@ -1610,6 +1646,92 @@ template NeoSupport ( )
+        public RemoveChannelResult removeChannel ( cstring channel )
+        {
+            auto task = Task.getThis();
+            assert(task, "This method may only be called from inside a Task");
```
`assert(task !is null)`

nemanja-boric-sociomantic commented:
Is there a way to adapt fakedht to not allow removing channel if the channel is being mirrored (and add appropriate tests)?

gavin-norman-sociomantic commented:
>Is there a way to adapt fakedht to not allow removing channel if the channel is being mirrored (and add appropriate tests)?

Why?

nemanja-boric-sociomantic commented:
Because it's advertised as part of the `RemoveChannel`'s specification, not as the implementation detail:

>  +  * A channel that is being mirrored cannot be removed. 

I do admit it's a nitpick.

gavin-norman-sociomantic commented:
Ahh I see. This is actually outdated. See https://github.com/sociomantic-tsunami/dhtproto/issues/39.

Commit:
Remove meaningless `static` from structs

gavin-norman-sociomantic commented:
Updated.

gavin-norman-sociomantic commented:
Updated with `handleSupportedCodes` simplification.
